src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -949,6 +949,117 @@ protected static (Socket, Socket) CreateConnectedSocketPair()
                 return (client, server);
             }
         }
+
+        [Fact]
+        [PlatformSpecific(~TestPlatforms.OSX)] // Not supported on OSX.
+        public async Task UdpReceiveGetsCanceledByDispose()
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            socket.BindToAnonymousPort(IPAddress.Loopback);
+
+            Task receiveTask = Task.Run(async () =>
+            {
+                await ReceiveAsync(socket, new ArraySegment<byte>(new byte[1]));
+            });
+
+            Task disposeTask = Task.Run(async () =>
+            {
+                // Wait a little so the receive is started.
+                await Task.Delay(100);
+
+                socket.Dispose();
+            });
+
+            Task timeoutTask = Task.Delay(30000);
+
+            Assert.NotSame(timeoutTask, await Task.WhenAny(disposeTask, receiveTask, timeoutTask));
+
+            await disposeTask;
+
+            try
+            {
+                await receiveTask;
+            }
+            catch (SocketException)
+            {}
+            catch (ObjectDisposedException)
+            {}
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        [PlatformSpecific(~TestPlatforms.OSX)] // Not supported on OSX.
```

Comment:
why not? This should work on OSX if we call shutdown()

Reply:
peerObservedConnectionAborted isn't true if we call shutdown.

Reply:
maybe we can update the test to handle that as well.(or ignore that part)  The important part IMHO is that we do unblock stuck socket.

Reply:
> IMHO is that we do unblock stuck socket.

We have a number of options.
- FIN on Unix
- FIN everywhere (including Windows)
- FIN where RST is not supported (OSX, ...)